File: servers/nextjs/lib/gcsStorage.ts
import { Storage } from "@google-cloud/storage";
import path from "path";

const getBucketName = () => process.env.GCS_BUCKET || "";
const getPrefix = () => process.env.GCS_PREFIX || "app_data";

const normalizeObjectName = (value: string) =>
  value.replace(/\\/g, "/").replace(/^\/+/, "");

export const uploadBufferToGcs = async (
  buffer: Buffer | Uint8Array,
  relativePath: string,
  contentType?: string
): Promise<string | null> => {
  const bucketName = getBucketName();
  if (!bucketName) {
    return null;
  }

  const prefix = getPrefix();
  const objectName = normalizeObjectName(path.posix.join(prefix, relativePath));
  const storage = new Storage();
  const file = storage.bucket(bucketName).file(objectName);

  await file.save(buffer as Buffer, {
    resumable: false,
    contentType: contentType || undefined,
  });
  await file.makePublic();

  return `https://storage.googleapis.com/${bucketName}/${objectName}`;
};

File: servers/fastapi/utils/gcs_storage.py
import os
from typing import Optional

from google.cloud import storage

from utils.get_env import get_app_data_directory_env, get_gcs_bucket_env, get_gcs_prefix_env


def _normalize_relative_path(path_value: str) -> str:
    return path_value.replace(os.sep, "/").lstrip("/")


def _derive_relative_path(local_path: str) -> str:
    app_data_dir = get_app_data_directory_env() or "/tmp/presenton"
    if local_path.startswith(app_data_dir):
        relative = local_path[len(app_data_dir) :]
    else:
        relative = os.path.basename(local_path)
    return _normalize_relative_path(relative)


def maybe_upload_file_to_gcs(
    local_path: str, relative_path: Optional[str] = None
) -> Optional[str]:
    bucket_name = get_gcs_bucket_env()
    if not bucket_name:
        return None

    prefix = get_gcs_prefix_env() or "app_data"
    relative = relative_path or _derive_relative_path(local_path)
    blob_name = _normalize_relative_path(os.path.join(prefix, relative))

    client = storage.Client()
    bucket = client.bucket(bucket_name)
    blob = bucket.blob(blob_name)
    blob.upload_from_filename(local_path)
    blob.make_public()

    # Object is now publicly readable.
    return f"https://storage.googleapis.com/{bucket_name}/{blob_name}"